specs/agents/metrics-otel.md

```diff
@@ -0,0 +1,181 @@
+# OpenTelemetry Metrics API Support
+
+Agents SHOULD support collecting custom metrics via the OpenTelemetry Metrics API.
+This SHOULD be done by supplying an OpenTelemetry Metrics SDK compatible [MetricExporter](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#metricexporter). If a pull based approach is preferred, a [MetricReader](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#metricreader) can be used instead. This `MetricExporter/MetricReader` SHOULD transform the received metrics data to IntakeV2 metricsets and send them via the existing reporting mechanism.
+
+## Metric naming
+
+OpenTelemetry metric names are allowed to contain dots. This can lead to mapping problems in elasticsearch if for example metrics with the names `foo.bar` and `foo` are both ingested.
+Due to the nested object representation within metricsets, `foo` would need to be mapped as both an object and a number at the same time, which is not possible.
+
+To avoid such conflicts, agents MUST replace all dots in metric names with `_` upon export if the `dedot_custom_metrics` configuration option is set to `true`:
```

Comment:
I may be entirely wrong here due to lack of context, but it seems like the renaming of metrics here happens on the elastic agents (running on the service), and are sent already renamed to the OTLP provider (presumably apm-server).

Wouldn't this make things harder for folks not using the agent, but the OpenTelemetry libraries, where they would have to ensure their metric names by themselves?
Shouldn't this renaming happen on the OTLP endpoint (the apm server)?

Reply:
@dmathieu yes, there is some important context here - let me try to summarize.

In short, what we specify here is sending OTel metrics from our existing APM Agents. Those APM Agents don't use OTLP, they have their own API to communicate with APM Server - this is the [`IntakeV2`](https://github.com/elastic/apm-server/tree/main/docs/spec/v2) mentioned in this spec. So APM Server supports OTLP and this IntakeV2 (and another one for RUM, but that's not relevant here). Plain OTel agents can send data with OTLP - and that is indeed handled in APM Server, but this scenario is for existing Elastic APM Users who want to move into OTel, but use our Elastic APM Agents - and in that case we bridge data already in the APM Agents for now with this spec.

> the renaming of metrics here happens on the elastic agents (running on the service)

That's correct

> to the OTLP provider (presumably apm-server). 

So as mentioned, there is no OTLP here - we send it via json with our IntakeV2 API.

> Wouldn't this make things harder for folks not using the agent, but the OpenTelemetry libraries, where they would have to ensure their metric names by themselves?

We try to have the same end result with this bridge and sending directly via OTLP. Not sure we'll get there in the first iteration, but long term sending via this bridge and via OTLP should have the same end result.

> Shouldn't this renaming happen on the OTLP endpoint (the apm server)?

Ideally yes, but the APM Server team is super overloaded - we'd like to make progress now and we decided to do this in the agents.

Reply:
Thank you for your detailed answer, that definitely makes more sense.
I wonder if adding a note stating that agents having to do this is only necessary until the backend does it would make things clearer?